src/db/mysql/batch.rs

```diff
@@ -0,0 +1,272 @@
+#![allow(proc_macro_derive_resolution_fallback)]
+
+use diesel::{
+    self, dsl::sql, insert_into, sql_types::Integer, update, ExpressionMethods, OptionalExtension,
+    QueryDsl, RunQueryDsl, TextExpressionMethods,
+};
+use serde_json;
+
+use super::{
+    models::{MysqlDb, Result},
+    schema::batches,
+};
+use db::{params, results, DbError};
+
+/// Rough guesstimate of the maximum reasonable life span of a batch.
+pub const BATCH_LIFETIME: i64 = 2 * 60 * 60 * 1000; // 2 hours, in milliseconds
+
+pub fn create(db: &MysqlDb, params: params::CreateBatch) -> Result<results::CreateBatch> {
+    let user_id = params.user_id.legacy_id as i32;
+    let collection_id = db.get_collection_id(&params.collection)?;
+    let timestamp = db.timestamp().as_i64();
+    // Sync timestamps are in seconds and quantized to two decimal places, so
+    // when we convert one to a bigint in milliseconds, the final digit is
+    // always zero. But we want to use the lower digits of the batchid for
+    // sharding writes via (batchid % num_tables), and leaving it as zero would
+    // skew the sharding distribution.
+    //
+    // So we mix in the lowest digit of the uid to improve the distribution
+    // while still letting us treat these ids as millisecond timestamps.  It's
+    // yuck, but it works and it keeps the weirdness contained to this single
+    // line of code.
+    let id = timestamp + (user_id % 10) as i64;
+    let bsos = bsos_to_batch_string(&params.bsos)?;
+    insert_into(batches::table)
+        .values((
+            batches::user_id.eq(&user_id),
+            batches::collection_id.eq(&collection_id),
+            batches::id.eq(&id),
+            batches::bsos.eq(&bsos),
+            batches::expiry.eq(timestamp + BATCH_LIFETIME),
+        )).execute(&db.conn)?;
+    Ok(id)
+}
+
+pub fn validate(db: &MysqlDb, params: params::ValidateBatch) -> Result<bool> {
+    let user_id = params.user_id.legacy_id as i32;
+    let collection_id = db.get_collection_id(&params.collection)?;
+    let exists = batches::table
+        .select(sql::<Integer>("1"))
+        .filter(batches::user_id.eq(&user_id))
+        .filter(batches::collection_id.eq(&collection_id))
+        .filter(batches::id.eq(&params.id))
+        .get_result::<i32>(&db.conn)
+        .optional()?;
+    Ok(exists.is_some())
+}
+
+pub fn append(db: &MysqlDb, params: params::AppendToBatch) -> Result<()> {
+    // XXX: this is possibly an upsert
```

Comment:
Not sure I agree about an upsert here, you should always know whether you have a new batch or an existing batch.